pkg/webhook/subnet/validator.go

```diff
@@ -0,0 +1,151 @@
+package subnet
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"strings"
+
+	ctlcniv1 "github.com/harvester/harvester/pkg/generated/controllers/k8s.cni.cncf.io/v1"
+	"github.com/harvester/webhook/pkg/server/admission"
+	kubeovnv1 "github.com/kubeovn/kube-ovn/pkg/apis/kubeovn/v1"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	kubeovnnetworkv1 "github.com/harvester/harvester-network-controller/pkg/generated/controllers/kubeovn.io/v1"
+	"github.com/harvester/harvester-network-controller/pkg/utils"
+)
+
+type Validator struct {
+	admission.DefaultValidator
+	nadCache ctlcniv1.NetworkAttachmentDefinitionCache
+	vpcCache kubeovnnetworkv1.VpcCache
+}
+
+var _ admission.Validator = &Validator{}
+
+func NewSubnetValidator(nadCache ctlcniv1.NetworkAttachmentDefinitionCache, vpcCache kubeovnnetworkv1.VpcCache) *Validator {
+	return &Validator{
+		nadCache: nadCache,
+		vpcCache: vpcCache,
+	}
+}
+
+func (v *Validator) Create(_ *admission.Request, newObj runtime.Object) error {
+	subnet := newObj.(*kubeovnv1.Subnet)
+
+	subnetSpec := subnet.Spec
+
+	if subnetSpec.Provider == "" {
+		return fmt.Errorf("provider is empty for subnet %s in %s", subnet.Name, subnet.Namespace)
+	}
+
+	if subnetSpec.CIDRBlock == "" || subnetSpec.Gateway == "" {
+		return fmt.Errorf("subnet cidr/gw is empty for subnet %s in %s", subnet.Name, subnet.Namespace)
+	}
```

Comment:
Do we want to allow empty-gateway subnets? The if statement and the error message seem not aligned.